internal/requests/requests.go

```diff
@@ -0,0 +1,63 @@
+package requests
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/internal/perms"
+)
+
+// ContextRequestInforation is a type used solely for context keys -- see the
+// variable below
+type ContextRequestInformation struct{}
+
+// ContextRequestInformationKey is a value to keep linters from complaining
+// about clashing identifiers
+var ContextRequestInformationKey ContextRequestInformation
+
+// RequestContext is used to propoagate request information. It can be updated
+// at various points, e.g. UserId would be updated via the result of
+// auth.Verify.
+//
+// It serves as a struct to gather information we learn about the request as it
+// goes along. This is distinct from options handling; none of the fields here
+// are "optional" and should all eventually be populated as the request moves
+// through the system.
+type RequestContext struct {
+	// Method is the method of the request
+	Method string
+
+	// Path is the path of the request
+	Path string
+
+	// UserId contains the final discovered user ID
+	UserId string
+
+	// OutputFields is the set of fields authorized for output for the
+	// authorized action, if not the default
+	OutputFields perms.OutputFieldsMap
+}
+
+func NewRequestContext(parent context.Context) context.Context {
+	return context.WithValue(parent, ContextRequestInformationKey, &RequestContext{})
+}
+
+// RequestContextFromCtx pulls out RequestContext and returns it. It will always
```

Comment:
The downside to this design is: the caller may believe that the data they add to RequestContext is propagated and there's no way to tell if it is or isn't.

Reply:
you could keep it very similar and just return a bool if the RequestContext was found.  Then it's up to the caller to determine the proper thing to do.